internal/dag/gatewayapi_processor.go

```diff
@@ -1425,6 +1430,19 @@ func (p *GatewayAPIProcessor) computeHTTPRouteForListener(
 				timeoutPolicy)
 		}
 
+		routes = append(routes, routesPerRule...)
+	}
+
+	// check all the routes at once in case there is conflict
+	if p.hasConflictRoute(listener, hosts, routes) {
+		// skip adding the route to svhost or vhost since it's marked as conflict route
```

Comment:
I think we might want to reconsider what happens when a route has at least one rule that conflicts with another but has some rules that are still valid (so as not to drop *all* traffic for this route)

we may want to not drop all the route rules and use the `PartiallyInvalid: True` Condition: https://github.com/kubernetes-sigs/gateway-api/blob/72f679c7c67228f5b4340e0c2f50cc04b36945b1/apis/v1/shared_types.go#L435

cc @projectcontour/maintainers @projectcontour/contour-reviewers

Reply:
Yeah i remember i read this similar discussion somewhere, do we expect rest of the route to be valid if just some of them returns 500

Reply:
But the route object we are  trying to drop here is already single route rule I assume? the matching criteria is constructed by `	PathMatchCondition&&HeaderMatchConditions&&QueryParamMatchConditions` under this route

Or do I miss anything?

Reply:
~~But the route object we are  trying to drop here is already single route rule I assume? the matching criteria is constructed by `	PathMatchCondition&&HeaderMatchConditions&&QueryParamMatchConditions` under this route~~

Oh yeah the logic I have now is to mark Route as invalid and not adding all routes under it to hosts.

Do we want to change it to partiallyInvalid? cc @skriss . I think it make snse

Reply:
Ok create an issue here https://github.com/projectcontour/contour/issues/6227
Also let me update the logic then

Reply:
```
	// 1) Drop Rule(s): With this approach, implementations will drop the
	//    invalid Route Rule(s) until they are fully valid again. The message
	//    for this condition MUST start with the prefix "Dropped Rule" and
	//    include information about which Rules have been dropped. In this
	//    state, the "Accepted" condition MUST be set to "True" with the latest
	//    generation of the resource.
```

https://github.com/kubernetes-sigs/gateway-api/blob/72f679c7c67228f5b4340e0c2f50cc04b36945b1/apis/v1/shared_types.go#L408C1-L413C35

The question i have here is it says about dropping `Rule`, so do we consider conflict at Rule level? If there is one route under it has conflict, we drop the rule?

Reply:
It sounds like we might be switching the definition of "route" and "rule" here?

In the upstream spec "route" should refer to the whole k8s resource (HTTPRoute) vs. "rule" is https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteRule

So if there is conflict with a single "rule" under a "route", only that rule would be dropped and other parts of the route should still work

Reply:
I was talking about this [Route object](https://github.com/lubronzhan/contour/blob/topic/lubron/fix-3608/internal/dag/dag.go#L277-L377). 
Should I call it cluster route? It's fetched from [here](https://github.com/projectcontour/contour/blob/main/internal/dag/gatewayapi_processor.go#L1396-L1427) 

What's the difference between this Route to the HttpRoute object?

Reply:
ah gotcha, yeah the `dag.Route` type more matches up the the `HTTPRouteRule` type (the `dag.VirtualHosts` and the `dag.Routes` under them basically come from a `HTTPRoute` and the Gateway `Listeners`, including the hostnames configured on the Route and Gateway)

so if we're processing a `HTTPRouteRule`+hostname (from the `HTTPRoute` and/or `Listener`) that has match conditions that conflict with an existing `dag.VirtualHost`+`dag.Route` that has already been added to the DAG (and therefore a `HTTPRouteRule` from another `HTTPRoute`) then we should not add it, set status, etc.

one complication I forgot to mention is that we've sort of implicitly allowed the processors that run later in the DAG build process to overwrite `dag.Routes` etc. from older ones (Gateway API can overwrite HTTPProxy can overwrite Ingress), if we make this change without any awareness of what can overwrite what, then we're kinda reversing this for the Gateway API processor (it won't be able to have precedence over other kinds)

Reply:
was thinking something like this might work:

```diff
diff --git a/internal/dag/dag.go b/internal/dag/dag.go
index b2bf912c..7710483d 100644
--- a/internal/dag/dag.go
+++ b/internal/dag/dag.go
@@ -769,6 +769,21 @@ func (v *VirtualHost) HasConflictRoute(route *Route) bool {
 	return false
 }
 
+func (v *VirtualHost) AddRouteIfNoConflict(route *Route) bool {
+	if v.Routes == nil {
+		v.Routes = make(map[string]*Route)
+	}
+	existingRoute, found := v.Routes[conditionsToString(route)]
+	if found {
+		return false
+	}
+	if existingRoute.Kind != route.Kind {
+		v.Routes[conditionsToString(route)] = route
+		return true
+	}
+	return false
+}
+
 func conditionsToString(r *Route) string {
 	s := []string{r.PathMatchCondition.String()}
 	for _, cond := range r.HeaderMatchConditions {
diff --git a/internal/dag/gatewayapi_processor.go b/internal/dag/gatewayapi_processor.go
index de518151..e2118fb5 100644
--- a/internal/dag/gatewayapi_processor.go
+++ b/internal/dag/gatewayapi_processor.go
@@ -1434,27 +1434,27 @@ func (p *GatewayAPIProcessor) computeHTTPRouteForListener(
 	}
 
 	// check all the routes at once in case there is conflict
-	if p.hasConflictRoute(listener, hosts, routes) {
-		// skip adding the route to svhost or vhost since it's marked as conflict route
-		routeAccessor.AddCondition(
-			gatewayapi_v1.RouteConditionAccepted,
-			meta_v1.ConditionFalse,
-			status.ReasonRouteRuleMatchConflict,
-			"HTTPRoute's Match has conflict with other HTTPRoute's Match",
-		)
-	} else {
-		// Add each route to the relevant vhost(s)/svhosts(s).
-		for host := range hosts {
-			for _, route := range routes {
-				switch {
-				case listener.tlsSecret != nil:
-					svhost := p.dag.EnsureSecureVirtualHost(listener.dagListenerName, host)
-					svhost.Secret = listener.tlsSecret
-					svhost.AddRoute(route)
-				default:
-					vhost := p.dag.EnsureVirtualHost(listener.dagListenerName, host)
-					vhost.AddRoute(route)
-				}
+	// Add each route to the relevant vhost(s)/svhosts(s).
+	for host := range hosts {
+		for _, route := range routes {
+			routeAdded := false
+			switch {
+			case listener.tlsSecret != nil:
+				svhost := p.dag.EnsureSecureVirtualHost(listener.dagListenerName, host)
+				svhost.Secret = listener.tlsSecret
+				routeAdded = svhost.AddRouteIfNoConflict(route)
+			default:
+				vhost := p.dag.EnsureVirtualHost(listener.dagListenerName, host)
+				routeAdded = vhost.AddRouteIfNoConflict(route)
+			}
+			if !routeAdded {
+				// skip adding the route to svhost or vhost since it's marked as conflict route
+				routeAccessor.AddCondition(
+					gatewayapi_v1.RouteConditionAccepted,
+					meta_v1.ConditionFalse,
+					status.ReasonRouteRuleMatchConflict,
+					"HTTPRoute's Match has conflict with other HTTPRoute's Match",
+				)
 			}
 		}
 	}
```

Reply:
Thanks for the detail, I checked the code again, one `dag.Route` is generated from one of the `Match` under `HTTPRouteRule`, so one rule could have several matches, thus several `dag.Route`, if we find single `Match` or `dag.Route`, under a `HTTPRouteRule` has conflict with other `dag.Route` with other `dag.Route`, then we should drop the whole `HTTPRouteRule`, not just the single `dag.Route` I assume.

Now I understand more about the relationship between `Route`, `dag.Route` and `Rule`, thanks

Reply:
MM not sure how do we indicate which rule has been dropped, since it doesn't have name under it, and it contains many matches. I guess I'll skip adding that info to the condition for now.
```
"Dropped Rule" and
	//    include information about which Rules have been dropped. 

```